src/crimson/common/gated.h

```diff
@@ -52,4 +68,43 @@ class Gated {
   seastar::gate pending_dispatch;
 };
 
-}// namespace crimson::common
+class gate_per_shard : public Gated {
+ public:
+  gate_per_shard() : gates(seastar::smp::count) {}
+  explicit gate_per_shard(size_t shard_count) : gates(shard_count) {}
+  gate_per_shard(const gate_per_shard&) = delete;
+  gate_per_shard& operator=(const gate_per_shard&) = delete;
+  gate_per_shard(gate_per_shard&&) = default;
+  gate_per_shard& operator=(gate_per_shard&&) = default;
+  ~gate_per_shard() override = default;
+
+  template <typename Func, typename T>
+  inline auto dispatch_in_background(const char* what, T& who, Func&& func) {
+    return dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func, typename T>
+  inline auto dispatch(const char* what, T& who, Func&& func) {
+    return gates[seastar::this_shard_id()].dispatch(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func>
+  auto submit_with_gate(const char* what, Func&& func) {
+    return gates[seastar::this_shard_id()].submit_with_gate(what, std::forward<Func>(func));
+  }
+
+  bool is_closed() const {
+    return gates[seastar::this_shard_id()].is_closed();
+  }
+
+  seastar::future<> close_all() {
+    return seastar::parallel_for_each(gates, [](Gated& gated) {
+      return gated.close();
```

Comment:
The gate needs to be closed in the specific shard where it is used, otherwise the same old issue could happen, right?

Reply:
I don't think so, as soon as you created the gate, you need to close it anyway, you can't close it more then one time, but you have to close it so the gate and holder will be destroyed later on and the debug gated will work and count the gates

Reply:
Yeah, the gates need to be closed.

I meant the specific gate should be closed in the shard where it is used in `dispatch()`. For example, `gates[4]` is used in `shard 4`, so it should be closed also in `shard 4`, because `close()` will access the internal `_count` which shouldn't be safe from a different shard.

Reply:
Correct, inside `parallel_for_each()`, refer to `seastar::smp::submit_to(core, fn)` about how to run logic in a specific shard.

Reply:
@NitzanMordhai, This comment was't addressed yet, right?

Reply:
@Matan-B now it is. took me some time to get it

Reply:
> Correct, inside `parallel_for_each()`, refer to `seastar::smp::submit_to(core, fn)` about how to run logic in a specific shard.

@cyx1231st please let me know if that's what you meant (I changed close_all)

Reply:
I think that you got it right with `submit_to` usage it `close_all`. Why did you drop `parallel_for_each` usage?

Reply:
In case the reason was to use the index of the loop as shard id, we can introduce `get_shard_id()` for Gated.

Reply:
@NitzanMordhai Yes, looks good to me now. Feel free to reach me in slack if my comment isn't clear in the future.

Regarding to `parallel_for_each`, and to make code cleaner, probably the following seastar document is useful:
```
/// \note parallel_for_each() schedules all invocations of \c func on the
///       current shard. If you want to run a function on all shards in parallel,
///       have a look at \ref smp::invoke_on_all() instead.
```